Lib/fontTools/ttLib/tables/_k_e_r_n.py

```diff
@@ -115,50 +132,95 @@ def decompile(self, data, ttFont):
 		searchRange, entrySelector, rangeShift = getSearchRange(nPairs, 6)
-		data = struct.pack(">HHHH", nPairs, searchRange, entrySelector, rangeShift)
+		data = struct.pack(
+			">HHHH", nPairs, searchRange, entrySelector, rangeShift)
 
 		# yeehee! (I mean, turn names into indices)
 		try:
 			reverseOrder = ttFont.getReverseGlyphMap()
-			kernTable = sorted((reverseOrder[left], reverseOrder[right], value) for ((left,right),value) in self.kernTable.items())
+			kernTable = sorted(
+				(reverseOrder[left], reverseOrder[right], value)
+				for ((left, right), value) in self.kernTable.items())
 		except KeyError:
 			# Slower, but will not throw KeyError on invalid glyph id.
 			getGlyphID = ttFont.getGlyphID
-			kernTable = sorted((getGlyphID(left), getGlyphID(right), value) for ((left,right),value) in self.kernTable.items())
+			kernTable = sorted(
+				(getGlyphID(left), getGlyphID(right), value)
+				for ((left, right), value) in self.kernTable.items())
 
 		for left, right, value in kernTable:
 			data = data + struct.pack(">HHh", left, right, value)
-		return struct.pack(">HHH", self.version, len(data) + 6, self.coverage) + data
+		# ensure mask bits 8-15 (subtable format) are set to 0
+		self.coverage &= ~0xFF
+
+		if not self.apple:
+			version = 0
+			length = len(data) + 6
+			header = struct.pack(">HHH", version, length, self.coverage)
+		else:
+			if self.tupleIndex is None:
+				# sensible default when compiling a TTX from an old fonttools
+				# or when inserting a Windows-style format 0 subtable into an
+				# Apple version=1.0 kern table
+				log.warning("'tupleIndex' is None; default to 0")
+				self.tupleIndex = 0
+			length = len(data) + 8
+			header = struct.pack(">LHH", length, self.coverage, self.tupleIndex)
+		return header + data
 
 	def toXML(self, writer, ttFont):
-		writer.begintag("kernsubtable", coverage=self.coverage, format=0)
+		attrs = dict(coverage=self.coverage, format=self.format)
+		if self.apple:
+			if self.tupleIndex is None:
+				log.warning("'tupleIndex' is None; default to 0")
+				attrs["tupleIndex"] = 0
+			else:
+				attrs["tupleIndex"] = self.tupleIndex
+		writer.begintag("kernsubtable", **attrs)
 		writer.newline()
 		items = sorted(self.kernTable.items())
 		for (left, right), value in items:
 			writer.simpletag("pair", [
-					("l", left),
-					("r", right),
-					("v", value)
-					])
+				("l", left),
+				("r", right),
+				("v", value)
+			])
 			writer.newline()
 		writer.endtag("kernsubtable")
 		writer.newline()
 
 	def fromXML(self, name, attrs, content, ttFont):
 		self.coverage = safeEval(attrs["coverage"])
-		self.version = safeEval(attrs["format"])
+		subtableFormat = safeEval(attrs["format"])
+		if self.apple:
+			if "tupleIndex" in attrs:
+				self.tupleIndex = safeEval(attrs["tupleIndex"])
+			else:
+				# previous fontTools versions didn't export tupleIndex
+				log.warning(
+					"Apple kern subtable is missing 'tupleIndex' attribute")
+				self.tupleIndex = None
```

Comment:
here I decided to set tupleIndex to None when it's missing from a kern TTX file; if I had set it to 0, it wouldn't be possible to distinguish from a kern TTX file with a missing tupleIndex attribute or one with tupleIndex=0.
The value still defaults to 0 when the imported font is compiled or dumped again to XML.